### ISSUE TITLE
Two workflows are running at the same time despite of semaphore

### ISSUE THREAD

**Issue by yaaraze** [argoproj/argo-workflows#10286]
### Pre-requisites

- [X] I have double-checked my configuration
- [X] I can confirm the issues exists when I tested with `:latest`
- [ ] I'd like to contribute the fix myself (see [contributing guide](https://github.com/argoproj/argo-workflows/blob/master/docs/CONTRIBUTING.md))

### What happened/what you expected to happen?

Hi, we’re using argo-workflows for two years now and recently we’ve experienced several times now that **two workflows are running at the same time even though they use the same semaphore**. 
Last time it just happened in the middle of the run after a few hours. No weird logs were shown, it looked like the second workflow was waiting on the semaphore and then just started running. Simultaneously we have other workflows running with other semaphores.

The only thing I can think of that we’ve changed recently is that we started using exit handlers.

Attaching the workflow file (I removed any env/args/labels/artifacts that I can't share).
Obviously this is a statistical bug so I can't reproduce it, but it happened three in the last two months (that I could notice) so I think it's pretty severe.

### Version

v3.2.4

### Paste a small workflow that reproduces the issue. We must be able to run the workflow; don't enter a workflows that uses private images.

```YAML
{
  "apiVersion": "argoproj.io/v1alpha1",
  "kind": "Workflow",
  "metadata": {
    "generateName": "argo-tenant-0x2-",
    "namespace": "argo"
  },
  "spec": {
    "onExit": "exit-handler",
    "ttlStrategy": {
      "secondsAfterSuccess": 1209600,
      "secondsAfterFailure": 1209600
    },
    "entrypoint": "main",
    "synchronization": {
      "semaphore": {
        "configMapKeyRef": {
          "name": "tenant-2-workflow-semaphore",
          "key": "workflow"
        }
      }
    },
    "templates": [
      {
        "name": "main",
        "steps": [
          [
            {
              "name": "wrapper-start-step",
              "template": "wrapper-start-template"
            }
          ],
          [
            {
              "name": "rabbitmq",
              "template": "rabbitmq-template"
            }
          ],
          [
            {
              "name": "first-step-initiator-step",
              "template": "first-step-initiator-template"
            }
          ],
          [
            {
              "name": "first-step-monitor-step",
              "template": "first-step-monitor-template"
            },
            {
              "name": "first-step-worker-step",
              "template": "first-step-worker-template",
              "withSequence": {
                "count": "5"
              }
            }
          ],
          [
            {
              "name": "wrapper-finish-step",
              "template": "wrapper-finish-template"
            }
          ]
        ]
      },
      {
        "name": "exit-handler",
        "steps": [
          [
            {
              "name": "wrapper-fail-step",
              "template": "wrapper-fail-template",
              "when": "{{workflow.status}} != Succeeded"
            }
          ]
        ]
      },
      {
        "name": "wrapper-start-template",
        "container": {
          "image": "hidden/wrapper:latest",
          "imagePullPolicy": "IfNotPresent",
          "command": ["python", "-m", "wrapper"],
          "args": ["--run-phase", "start"]
        },
        "serviceAccountName": "hidden-service-account"
      },
      {
        "name": "wrapper-fail-template",
        "container": {
          "image": "hidden/wrapper:latest",
          "imagePullPolicy": "IfNotPresent",
          "command": ["python", "-m", "wrapper"],
          "args": ["--run-phase", "fail"]
        },
        "serviceAccountName": "hidden-service-account"
      },
      {
        "name": "first-step-initiator-template",
        "container": {
          "image": "hidden/first-step:latest",
          "imagePullPolicy": "IfNotPresent",
          "command": ["python", "-m", "first-step"],
          "args": ["--runner", "initiator"]
        },
        "serviceAccountName": "hidden-service-account"
      },
      {
        "name": "first-step-worker-template",
        "container": {
          "image": "hidden/first-step:latest",
          "imagePullPolicy": "IfNotPresent",
          "command": ["python", "-m", "first-step"],
          "args": ["--runner", "worker"]
        },
        "serviceAccountName": "hidden-service-account",
        "retryStrategy": {
          "limit": "3"
        }
      },
      {
        "name": "first-step-monitor-template",
        "container": {
          "image": "hidden/first-step:latest",
          "imagePullPolicy": "IfNotPresent",
          "command": ["python", "-m", "first-step"],
          "args": ["--runner", "monitor"]
        },
        "serviceAccountName": "hidden-service-account"
      },
      {
        "name": "rabbitmq-template",
        "container": {
          "image": "rabbitmq:3.11.2-alpine",
          "name": "rabbitmq",
          "ports": [
            {
              "containerPort": 5672
            }
          ]
        },
        "daemon": true
      },
      {
        "name": "wrapper-finish-template",
        "container": {
          "image": "hidden/wrapper:latest",
          "imagePullPolicy": "IfNotPresent",
          "command": ["python", "-m", "wrapper"],
          "args": ["--run-phase", "finish"]
        },
        "serviceAccountName": "hidden-service-account"
      }
    ]
  }
}
```

### Logs from the workflow controller

```text
.
```

### Logs from in your workflow's wait container

```text
.
```

**Comment by sarabala1979** [comment]
@yaaraze can you try 3.3.10 or 3.4.4?

**Comment by yaaraze** [comment]
Hi,
It still happens in 3.4.4

**Comment by tahiraha** [comment]
I'm running into this now on 3.4.4

**Comment by tooptoop4** [comment]
🌮

**Comment by tooptoop4** [comment]
anyone got controller logs for this?

**Comment by Joibel** [comment]
Thanks for the report, and apologies for how long this sat.

The most likely cause of this has since been found and fixed. #11098 was a reproducible version of the same symptom (an extra workflow admitted past a workflow-level semaphore) and was root-caused to a variable-shadowing bug in the sync manager's startup path: when the controller restarted, the first recorded holder of each semaphore key was silently dropped, so a limit-1 semaphore came back up with no holders and immediately admitted the queued workflow while the original was still running. That matches what you saw - no errors, the running workflow unaffected, and the waiting workflow starting at an apparently arbitrary point hours in (whenever the controller pod restarted). The bug had been there since the original semaphore implementation, so it affected both 3.2.4 and 3.4.4.

It was fixed in #15609 and released in v3.7.11 and v4.0.2. Since then the startup path has been rewritten again: recorded holds are re-established and verified, a lock that cannot be rebuilt is poisoned rather than left empty for another workflow to take, and database-backed holds that the database no longer records cause the workflow to be failed instead of running on a lock it does not hold.

Closing on that basis. If you still see two workflows sharing a limit-1 semaphore on a supported release, please open a new issue with workflow-controller logs covering the moment the second workflow starts - without those we can't distinguish this from other causes.
